File: solutions/clojure/state-of-tic-tac-toe/1/src/state_of_tic_tac_toe.clj
(ns state-of-tic-tac-toe)

(defn get-counts [board char]
    (count (filter #(= % char)  (clojure.string/split (clojure.string/join board) #""))))

(defn verify-board [board]
    (let [naughts (get-counts board "O")
          crosses (get-counts board "X")
          empty (get-counts board " ")]
      (cond
          (not (= (+ naughts crosses empty) 9)) (throw (IllegalArgumentException. "Wrong board format"))
          (> naughts crosses) (throw (IllegalArgumentException. "Wrong turn order: O started"))
          (> crosses (inc naughts)) (throw (IllegalArgumentException. "Wrong turn order: X went twice"))
          :else true)))

(defn line-win? [board no player]
  (= (nth board no) (clojure.string/join (repeat 3 player)))) 

(defn count-line-wins [board player]
  (count (filter #(line-win? board % player) (list 0 1 2))))

(defn count-diagonal-wins [board player]
  (let [line1 (clojure.string/split (nth board 0) #"")
        line2 (clojure.string/split (nth board 1) #"")
        line3 (clojure.string/split (nth board 2) #"")]
  (+ (if (and  (= (first line1) player)
    (= (nth line2 1) player)
    (= (last line3) player)) 1 0)
     (if (and  (= (last line1) player)
    (= (nth line2 1) player)
    (= (first line3) player)) 1 0))))

(defn column-win? [board no player]
  (= 3 (count (filter #(= (str %) player) (list (nth (nth board 0) no)
  (nth (nth board 1) no)
  (nth (nth board 2) no) )))))

(defn count-column-wins [board player]
  (count (filter #(column-win? board % player) (list 0 1 2))))

(defn count-wins [board player]
  (+ (count-line-wins board player) (count-diagonal-wins board player) (count-column-wins board player)))


(defn gamestate
  "Returns the gamestate of a tic-tac-toe board."
  [board]
  ;; function body
  (when (verify-board board)
    (let [cross-wins (count-wins board "X")
          naught-wins (count-wins board "O")]
      (if (and (> cross-wins 0) (> naught-wins 0))
        (throw (IllegalArgumentException. "Impossible board: game should have ended after the game was won"))
        (cond 
           (or (> cross-wins 0) (> naught-wins 0)) :win
           (zero? (get-counts board " ")) :draw 
           :else :ongoing
           )))))
  

File: solutions/python/state-of-tic-tac-toe/1/state_of_tic_tac_toe.py
def count_cells_for(board, player):
    return len([c for c in "".join(board) if c==player])

def count_wins_for(board, player):
    win_string = player*3
    count = 0
    count += len([row for row in board if row==win_string])
    count += len([i for i in {0,1,2} if 
                  player==board[0][i]==board[1][i]==board[2][i]])
    if board[0][0] == board[1][1] == board[2][2]==player:
        count +=1
    if board[2][0] == board[1][1] == board[0][2]==player:
        count += 1
    return count

def gamestate(board):
    cross = 'X'
    naught = 'O'

    cc = count_cells_for(board, cross)
    cn = count_cells_for(board, naught)

    if cn>cc:
        raise ValueError("Wrong turn order: O started")
    elif cc-cn>1:
        raise ValueError("Wrong turn order: X went twice")
    state = ""
    wc = count_wins_for(board, cross)
    wn = count_wins_for(board, naught)

    if 0==wc and 0==wn:
        if cc+cn==9:
            state = "draw"
        else:
            state = "ongoing"
    elif (wc>0 and wn>0) or (wc>2 or wn>2):
        raise ValueError("Impossible board: game should have ended after the game was won")
    else:
        state = "win"
    return state